sw/device/lib/drivers/gpio/gpio.c

```diff
@@ -1,213 +1,624 @@
+/****************************************************************************/
+
+/****************************************************************************/
+/**                                                                        **/
+/*                      PROTOTYPES OF LOCAL FUNCTIONS                       */
+/**                                                                        **/
+/****************************************************************************/
+
+/**
+ * @brief do a bitwise operation on 32b register.
+ * @param reg the register pointer that bitwise operation is done on it.
+ * @param val the value should be written in part of reg w/o offset.
+ * @param mask the mask w/o offset.
+ * @param sel this the offset used for mask and value.
  */
-static uint32_t index_to_mask(uint32_t index) { return 1u << index; }
+static inline void setBitfield( uint32_t *reg, uint32_t  val, 
+    uint32_t mask, uint8_t sel )
+{
+    *reg           &= ~( mask << sel );
+    *reg           |= (val & mask) << sel;
+}
+
+/**
+ * @brief get some bitfield of a 32b reg.
+ * @param reg the register that bitwise field selects from.
+ * @param mask the mask w/o offset.
+ * @param sel this the offset used for mask and value.
+ */
+static inline uint32_t getBitfield( uint32_t reg, uint32_t  mask, uint8_t sel )
+{
+    uint32_t ret_val = 0;
+    ret_val = reg & ( mask << sel );
+    ret_val = ret_val >> sel;
+    return ret_val;
+}
+
+/****************************************************************************/
+/**                                                                        **/
+/*                           EXPORTED VARIABLES                             */
+/**                                                                        **/
+/****************************************************************************/
+
+/****************************************************************************/
+/**                                                                        **/
+/*                            GLOBAL VARIABLES                              */
+/**                                                                        **/
+/****************************************************************************/
+
+/****************************************************************************/
+/**                                                                        **/
+/*                           EXPORTED FUNCTIONS                             */
+/**                                                                        **/
+/****************************************************************************/
+
+gpio_result_t gpio_config (gpio_cfg_t cfg){
+    // /* check that passed arg is not empty*/
+    // if (cfg == NULL)
+    //     return GpioError;
+    /* check that pin is in acceptable range. */
+    if (cfg.pin > (MAX_PIN-1) || cfg.pin < 0)
+        return GpioError;
```

Comment:
in this function, there are different conditions that can give an error, so the user should know what is exactly the thing that failed. Thus, I recommend implementing some error codes besides Ok and Error to help tackling/knowing this

Reply:
Ok, the checkings I was thinking to do was "pin number" being correct and also a structure in the form gpio_config being passed. Some of the checking I'm not doing internally, for example if the gpio is set as output, enable sampling is never set for it.
However, if you think those two are not enough, let me know

Reply:
I added several checkings, and based on them, the error type is different:
- Given pin is not inside of acceptable range
- Given pin mode is not one of the defined modes in gpio_mode_t
- Given interrupt type is not one of the defined types in gpio_intr_type_t

Reply:
Anything to be returned than 0 (GpioOk) is an error.